### PR TITLE
include: Add support for C compilation

### LINKED ISSUE #4: [include] C Compiler Support
I was taking a look at compiling our upcoming [D3D11 renderer for FNA3D](https://github.com/FNA-XNA/FNA3D/pull/19) against dxvk-native, but ran into some issues with the C++-isms in the Windows headers. For compatibility, the headers will likely need to act more like the real deal and work with C applications.

The build support is already integrated and ready to go, so it's just a matter of making it compile (it probably won't run since we need a d3dcompiler):

```
git clone --recursive https://github.com/FNA-XNA/FNA3D
git clone https://github.com/Joshua-Ashton/dxvk-native
mkdir FNA3D/flibitBuild
cd FNA3D/flibitBuild
cmake .. -DBUILD_DXVK_NATIVE=ON
make
```

### PR DESCRIPTION
**Pull request by flibitijibibo**
This fixes building FNA3D with dxvk-native.

As a reminder, the FNA3D build:
```
git clone --recursive https://github.com/FNA-XNA/FNA3D
git clone https://github.com/Joshua-Ashton/dxvk-native
mkdir FNA3D/flibitBuild
cd FNA3D/flibitBuild
cmake .. -DBUILD_DXVK_NATIVE=ON
make
```

Test program (throw this in flibitBuild):
```
/* cc clear.c -I../include `sdl2-config --cflags --libs` libFNA3D.so */

#include <SDL.h>
#include <FNA3D.h>

int main(int argc, char **argv)
{
	SDL_Window *window;
	FNA3D_Device *device;
	FNA3D_PresentationParameters params;
	FNA3D_Vec4 color;
	SDL_Event evt;
	SDL_bool run = SDL_TRUE;

	SDL_Init(SDL_INIT_VIDEO);
	window = SDL_CreateWindow(
		"FNA3D dxvk-native Test",
		SDL_WINDOWPOS_CENTERED,
		SDL_WINDOWPOS_CENTERED,
		640,
		480,
		SDL_WINDOW_SHOWN | FNA3D_PrepareWindowAttributes()
	);
	params.backBufferWidth = 640;
	params.backBufferHeight = 480;
	params.backBufferFormat = FNA3D_SURFACEFORMAT_COLOR;
	params.multiSampleCount = 0;
	params.deviceWindowHandle = window;
	params.isFullScreen = 0;
	params.depthStencilFormat = FNA3D_DEPTHFORMAT_NONE;
	params.presentationInterval = FNA3D_PRESENTINTERVAL_DEFAULT;
	params.displayOrientation = FNA3D_DISPLAYORIENTATION_DEFAULT;
	params.renderTargetUsage = FNA3D_RENDERTARGETUSAGE_DISCARDCONTENTS;
	device = FNA3D_CreateDevice(&params, 0);
	color.x = 100 / 255.0f;
	color.y = 149 / 255.0f;
	color.z = 237 / 255.0f;
	color.w = 1.0f;

	while (run)
	{
		while (SDL_PollEvent(&evt) > 0)
		{
			if (evt.type == SDL_QUIT)
			{
				run = SDL_FALSE;
			}
		}

		FNA3D_Clear(
			device,
			FNA3D_CLEAROPTIONS_TARGET,
			&color,
			1.0f,
			0
		);
		FNA3D_SwapBuffers(device, NULL, NULL, window);
	}

	FNA3D_DestroyDevice(device);
	SDL_DestroyWindow(window);
	SDL_Quit();
	return 0;
}
```

Patch Notes:
- THE WARNINGS, OH GOD THE WARNINGS oh it's just -Wpedantic
- A good amount of this can be guarded with `__cplusplus`, but not quite all... it might be better to replicate _exactly_ how the official headers do things.
- I mostly improvised some of the interface-y bits from how I remember it working but there's no way it isn't wrong.

Fixes #4

### CODE REVIEW

**Comment by flibitijibibo** [comment]
Fixed the dxvk build, this is "ready" to review.

**Comment by flibitijibibo** [comment]
Figured out the last of the Win32 quirks, this is now _actually_ ready for review.

File: include/native/windows/windows_base.h [hunk 1]
```diff
@@ -173,55 +192,60 @@ constexpr BOOL FALSE = 0;
...
-constexpr HRESULT DXGI_ERROR_CANNOT_PROTECT_CONTENT        = 0x887A002A;
-constexpr HRESULT DXGI_ERROR_ACCESS_DENIED                 = 0x887A002B;
-constexpr HRESULT DXGI_ERROR_NAME_ALREADY_EXISTS           = 0x887A002C;
-constexpr HRESULT DXGI_ERROR_SDK_COMPONENT_MISSING         = 0x887A002D;
-
-constexpr UINT D3DENUM_WHQL_LEVEL = 2;
+#define S_OK     0
+#define S_FALSE  1
+
+#define E_INVALIDARG  0x80070057
```
Comment (Joshua-Ashton) [inline include/native/windows/windows_base.h:206]: These should probably `#define THING ((HRESULT)value)`

File: include/native/windows/windows_base.h [hunk 2]
```diff
@@ -265,8 +308,17 @@ constexpr UINT D3DENUM_WHQL_LEVEL = 2;
...
 inline bool FAILED(T hr) { return HRESULT(hr) < 0; }
 
 template <typename T>
-inline bool SUCCEEDED(T hr) { return !FAILED<T>(hr); }
\ No newline at end of file
+inline bool SUCCEEDED(T hr) { return !FAILED<T>(hr); }
+
+#else
+
+#define FAILED(hr) (hr < 0)
```
Comment (Joshua-Ashton) [inline include/native/windows/windows_base.h:321]: May as well just stick with this everywhere...

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

**Comment by flibitijibibo** [comment]
Pushed with both fixes applied! A notable change is that `FAILED`/`SUCCEEDED` needed to wrap `(hr)` for complete safety.

**Pull request merged**